src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,503 @@
+
+            // make sure the collection can be iterated.
+            if (!TryGetIterationElementType(
+                    containingType, collectionType.Type,
+                    ienumerableType, ienumeratorType,
+                    out var iterationType))
+            {
+                return;
+            }
+
+            // If the user uses the iteration variable for any other reason, we can't convert this.
+            var bodyStatements = GetBodyStatements(forStatement);
+            foreach (var statement in bodyStatements)
+            {
+                if (iterationVariableIsUsedForMoreThanCollectionIndex(statement))
+                {
+                    return;
+                }
+            }
+
+            // Looks good.  We can convert this.
+            context.RegisterRefactoring(new MyCodeAction(GetTitle(),
+                c => ConvertForToForEachAsync(
+                    document, forStatement, iterationVariable, collectionExpression,
+                    containingType, collectionType.Type, iterationType, c)));
+
+            // local functions
+            bool iterationVariableIsUsedForMoreThanCollectionIndex(SyntaxNode current)
+            {
+                if (syntaxFacts.IsIdentifierName(current))
+                {
+                    syntaxFacts.GetNameAndArityOfSimpleName(current, out var name, out _);
+                    if (name == iterationVariable.ValueText)
+                    {
+                        // found a reference.  make sure it's only used inside something like
+                        // list[i]
+
+                        if (!syntaxFacts.IsSimpleArgument(current.Parent) ||
+                            !syntaxFacts.IsElementAccessExpression(current.Parent.Parent.Parent))
+                        {
+                            // used in something other than accessing into a collection.
+                            // can't convert this for-loop.
+                            return true;
+                        }
+
+                        var arguments = syntaxFacts.GetArgumentsOfArgumentList(current.Parent.Parent);
+                        if (arguments.Count != 1)
+                        {
+                            // was used in a multi-dimensional indexing.  Can't conver this.
+                            return true;
+                        }
+
+                        var expr = syntaxFacts.GetExpressionOfElementAccessExpression(current.Parent.Parent.Parent);
+                        if (!syntaxFacts.AreEquivalent(expr, collectionExpression))
+                        {
+                            // was indexing into something other than the collection.
+                            // can't convert this for-loop.
+                            return true;
+                        }
+
+                        // this usage of the for-variable is fine.
+                    }
+                }
+
+                foreach (var child in current.ChildNodesAndTokens())
```

Comment:
we have ChildNodes too.

Reply:
ChildNodes allocated.  ChildNodesAndTokens does not.

Reply:
For recursive walks, the IDE prefers ChildNodesAndTokens to prevent that cost at each node it hits.